### PR TITLE
Change channel type in event module 

### PR DESCRIPTION
**Pull request by pjueon**
This is a fix for the issue #39. 
**(Summary: The channel type in event module should be changed because it is NOT always convertible to int.  )**
Tested it with `samples/test_all_apis.cpp`. 

@ShimmyShaman 
Could you review the code please?  

Major changes:
- The signature of `wait_for_edge`, `add_event_detect ` functions changed 
- The signature of the callback changed

Want to discuss:
-  Any better idea for the return value of the `wait_for_edge` function?
   - it returns 0 on timeout if the channel is int. 
   - it returns an empty string on timeout if the channel is string.

### CODE REVIEW

**Comment by pjueon** [comment]
I came up with some options for the return type of wait_for_edge function: 

Option 1. Use different types for each overloading function (now)

```cpp
// return the channel for which the edge was detected. 
//If a timeout occurred, it returns 0(if the channel is int) or an empty string(if the channel is string).
std::string wait_for_edge(const std::string& channel, Edge edge, unsigned long bounce_time = 0, unsigned long timeout = 0);
int wait_for_edge(int channel, Edge edge, unsigned long bounce_time = 0, unsigned long timeout = 0);

//...


// in the user code,
if (wait_for_edge(...) == channel)
// or
// for string channel
if(!wait_for_edge(...).empty())

// for int channel
if(wait_for_edge(...))
// ...
```

Option 2. create a GPIO::WaitResult just for it.

```cpp

class WaitResult
{
public:
   WaitResult(const std::string& channel) : _channel(channel){}
   
   bool is_event_detected() const { return !channel().empty(); };
   operator bool() const { return is_event_detected(); }

   // can get the channel info
   const std::string& channel() const { return _channel; }

private:
   std::string _channel;
};

// both versions return GPIO::WaitResult object
WaitResult wait_for_edge(const std::string& channel, unsigned long bounce_time = 0, unsigned long timeout = 0);
WaitResult wait_for_edge(int channel, Edge edge, unsigned long bounce_time = 0, unsigned long timeout = 0);

// ...

// in the user code, 
if(wait_for_edge( .... ).is_event_detected())
// or just
if(wait_for_edge( .... ))
// ...

//...
auto result = wait_for_edge( .... );
std::string channel = wait_for_edge.channel(); 
```

Option 3. always return std::string
```cpp
// return empty string on timeout
 std::string wait_for_edge(const std::string& channel, Edge edge, unsigned long bounce_time = 0, unsigned long timeout = 0);
 std::string wait_for_edge(int channel, Edge edge, unsigned long bounce_time = 0, unsigned long timeout = 0);

// in the user code,
if (wait_for_edge(...) == channel)
// or
if(!wait_for_edge(...).empty())

```
In this way, the return types of 2 overloading functions are same. 

@ShimmyShaman 
Which one do you think the best? Personally, option 2 is my favorite.

**Comment by ShimmyShaman** [comment]
Option 2 is a nice solution.
All the code changes look functional and non-breaking. You've got my approval for what you've done.

_Won't click the merge button as I will let you implement your option and merge when you're ready._

**Review by ShimmyShaman — APPROVED** [review]

**Comment by pjueon** [comment]
Thank you for the review.
Alright, I'll implement option 2 and merge it when I get back home.

**Pull request merged**
